pkg/registry/generic/registry/store.go

```diff
@@ -268,6 +306,34 @@ func (e *Store) Update(ctx api.Context, obj runtime.Object) (runtime.Object, boo
 		UIDCopy := meta.UID
 		preconditions = &storage.Preconditions{UID: &UIDCopy}
 	}
+	// check if the update is removing all the object's finalizers. If so,
+	// further checks if the object's DeletionGracePeriodSeconds is 0. If so,
+	// delete the object.
+	existing := e.NewFunc()
+	delete := e.shouldDelete(ctx, key, obj, existing)
```

Comment:
I'll look more in the morning, but Update calling Delete seems odd (and the raciness is concerning). I really don't think we should be mixing update/finalize/delete concerns into a single function. 

However, if that is something we decide we're happy with, nothing stops you from doing the check for finalizers inside the GuaranteedUpdate loop (you need the old object anyway for resource version checks) and skipping the update by returning a marker error the outer function checks for and handles with a "didFinalize" case.